File: frontend/apps/ios/Pantopus/Features/Explore/ExploreMapViewModel.swift
//
//  ExploreMapViewModel.swift
//  Pantopus
//
//  A11.2 Explore — view-model for the cross-type discovery map. Holds the
//  loaded sample entities, the top type-toggle selection, the sort, the
//  applied filter criteria, and a single `selectedId` so pin taps and
//  rail-card highlights stay in sync. Filtering + clustering run locally
//  over the sample set (no network — backend removed from the repo).
//

import Foundation
import Observation

@Observable
@MainActor
public final class ExploreMapViewModel {
    public private(set) var state: ExploreMapState = .loading

    /// Active type-toggle selection. `nil` == "All".
    public private(set) var activeKind: ExploreKind?

    /// Sort applied to the sheet body (local, no refetch).
    public private(set) var activeSort: ExploreSort = .closest

    /// Current sheet stop. Driven by drag-release in the view.
    public var sheetStop: ExploreSheetStop = .standard

    /// Last user coordinate — drives the "you are here" disc.
    public private(set) var userCoordinate: UserCoordinate?

    /// Applied filter criteria. Surfaced so the view can seed the filter
    /// sheet + render the active-count badge.
    public private(set) var filters: ExploreFilterCriteria

    private let scenario: ExploreScenario
    private var allEntities: [ExploreEntity]
    /// Grid-bucket cluster radius (~0.005° ≈ 500 m at NYC latitude).
    private var clusterRadiusDegrees: Double = 0.005

    public init(scenario: ExploreScenario = .populated) {
        self.scenario = scenario
        allEntities = ExploreMapSampleData.entities(for: scenario)
        filters = ExploreMapSampleData.filters(for: scenario)
        userCoordinate = ExploreMapSampleData.center
    }

    // MARK: - Lifecycle

    public func load() async {
        switch scenario {
        case .loading:
            state = .loading
        case .error:
            state = .error(message: "Couldn't load the map.")
        case .populated, .empty:
            rebuild(selectedId: nil)
        }
    }

    public func refresh() async {
        await load()
    }

    // MARK: - Type toggle

    public func selectKind(_ kind: ExploreKind?) {
        guard kind != activeKind else { return }
        activeKind = kind
        rebuild(selectedId: keptSelection())
    }

    // MARK: - Sort

    public func selectSort(_ sort: ExploreSort) {
        guard sort != activeSort else { return }
        activeSort = sort
        rebuild(selectedId: currentSelectedId())
    }

    // MARK: - Filters

    /// Apply structured filters from the filter sheet. Re-projects pins +
    /// rail immediately and drops the selection if it no longer survives.
    public func applyFilters(_ criteria: ExploreFilterCriteria) {
        filters = criteria
        rebuild(selectedId: keptSelection())
    }

    /// Empty-state "Clear filters" — resets every dimension (including the
    /// type toggle) so the full neighborhood reappears.
    public func clearFilters() {
        filters = ExploreFilterCriteria()
        activeKind = nil
        rebuild(selectedId: nil)
    }

    /// Empty-state "Widen area" — opens the distance radius to the widest
    /// stop so neighbors a little further out surface, leaving the other
    /// dimensions intact.
    public func widenArea() {
        filters.distanceUpper = ExploreFilterCriteria.distanceStops[ExploreFilterCriteria.distanceDefaultIndex]
        rebuild(selectedId: currentSelectedId())
    }

    // MARK: - Selection

    /// Pin tap or rail-card tap — both feed the same selection so the
    /// pulse halo and the rail highlight stay tied.
    public func selectEntity(_ id: String?) {
        guard case let .loaded(loaded) = state else { return }
        state = .loaded(ExploreMapLoaded(
            entities: loaded.entities,
            markers: loaded.markers,
            userCoordinate: loaded.userCoordinate,
            selectedId: id
        ))
    }

    public func setSheetStop(_ stop: ExploreSheetStop) {
        sheetStop = stop
    }

    /// Reduce the cluster radius (camera zoom-in) and rebuild markers.
    public func setClusterRadius(_ radiusDegrees: Double) {
        let clamped = max(0.0005, min(radiusDegrees, 0.05))
        guard abs(clamped - clusterRadiusDegrees) > 1e-6 else { return }
        clusterRadiusDegrees = clamped
        if case let .loaded(loaded) = state {
            rebuild(selectedId: loaded.selectedId)
        }
    }

    // MARK: - Projection

    private func currentSelectedId() -> String? {
        if case let .loaded(loaded) = state { return loaded.selectedId }
        return nil
    }

    /// Keep the current selection only if it survives the new filter.
    private func keptSelection() -> String? {
        let prior = currentSelectedId()
        return filtered().contains { $0.id == prior } ? prior : nil
    }

    private func filtered() -> [ExploreEntity] {
        allEntities.filter { entity in
            if let activeKind, entity.kind != activeKind { return false }
            return filters.matches(entity)
        }
    }

    private func sorted(_ source: [ExploreEntity]) -> [ExploreEntity] {
        switch activeSort {
        case .closest:
            source.sorted { $0.distanceMiles < $1.distanceMiles }
        case .newest:
            // No timestamps in the stub set — preserve authored order
            // (hero entities first) as the "newest" projection.
            source
        }
    }

    private func rebuild(selectedId: String?) {
        let entities = sorted(filtered())
        state = .loaded(ExploreMapLoaded(
            entities: entities,
            markers: Self.cluster(entities: entities, radiusDegrees: clusterRadiusDegrees),
            userCoordinate: userCoordinate,
            selectedId: selectedId
        ))
    }

    // MARK: - Clustering

    /// Grid-bucket clusterer. Snaps each entity to a `radiusDegrees` grid;
    /// buckets of size 1 stay as `.entity` markers, buckets of size ≥2
    /// collapse into a `.cluster`. Order is stable (sorted by bucket key).
    static func cluster(entities: [ExploreEntity], radiusDegrees: Double) -> [ExploreMarker] {
        guard radiusDegrees > 0 else { return entities.map(ExploreMarker.entity) }
        var buckets: [String: [ExploreEntity]] = [:]
        for entity in entities {
            let key = bucketKey(latitude: entity.latitude, longitude: entity.longitude, radius: radiusDegrees)
            buckets[key, default: []].append(entity)
        }
        return buckets.keys.sorted().compactMap { key -> ExploreMarker? in
            guard let group = buckets[key] else { return nil }
            if group.count == 1 { return .entity(group[0]) }
            let lats = group.map(\.latitude)
            let lons = group.map(\.longitude)
            let centerLat = lats.reduce(0, +) / Double(group.count)
            let centerLon = lons.reduce(0, +) / Double(group.count)
            let representative = group
                .reduce(into: [:]) { (counts: inout [ExploreKind: Int], entity: ExploreEntity) in
                    counts[entity.kind, default: 0] += 1
                }
                .max { $0.value < $1.value }?.key ?? group[0].kind
            return .cluster(ExploreCluster(
                id: key,
                latitude: centerLat,
                longitude: centerLon,
                kind: representative,
                count: group.count,
                entityIds: group.map(\.id),
                minLatitude: lats.min() ?? centerLat,
                maxLatitude: lats.max() ?? centerLat,
                minLongitude: lons.min() ?? centerLon,
                maxLongitude: lons.max() ?? centerLon
            ))
        }
    }

    private static func bucketKey(latitude: Double, longitude: Double, radius: Double) -> String {
        let latBucket = Int((latitude / radius).rounded(.down))
        let lonBucket = Int((longitude / radius).rounded(.down))
        return "\(latBucket)_\(lonBucket)"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/explore/ExploreMapViewModel.kt
@file:Suppress("MagicNumber", "PackageNaming")

package app.pantopus.android.ui.screens.explore

import androidx.lifecycle.ViewModel
import app.pantopus.android.data.location.UserCoordinate
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject

/**
 * Backs the A11.2 Explore map. Holds the loaded sample entities, the top
 * type-toggle selection, the sort, the applied filter criteria, and a
 * single `selectedId` so pin taps and rail-card highlights stay in sync.
 * Filtering + clustering run locally over the sample set (no network —
 * backend removed from the repo). Mirrors `ExploreMapViewModel.swift`.
 */
@HiltViewModel
class ExploreMapViewModel
    @Inject
    constructor() : ViewModel() {
        private val _state = MutableStateFlow<ExploreMapUiState>(ExploreMapUiState.Loading)
        val state: StateFlow<ExploreMapUiState> = _state.asStateFlow()

        private val _activeKind = MutableStateFlow<ExploreKind?>(null)
        val activeKind: StateFlow<ExploreKind?> = _activeKind.asStateFlow()

        private val _activeSort = MutableStateFlow(ExploreSort.Closest)
        val activeSort: StateFlow<ExploreSort> = _activeSort.asStateFlow()

        private val _sheetStop = MutableStateFlow(ExploreSheetStop.Standard)
        val sheetStop: StateFlow<ExploreSheetStop> = _sheetStop.asStateFlow()

        private val _userCoordinate = MutableStateFlow<UserCoordinate?>(ExploreMapSampleData.center)
        val userCoordinate: StateFlow<UserCoordinate?> = _userCoordinate.asStateFlow()

        private val _filters = MutableStateFlow(ExploreFilterCriteria())
        val filters: StateFlow<ExploreFilterCriteria> = _filters.asStateFlow()

        private var scenario: ExploreScenario = ExploreScenario.Populated
        private var allEntities: List<ExploreEntity> = emptyList()
        private var clusterRadiusDegrees: Double = 0.005

        fun load(scenario: ExploreScenario = ExploreScenario.Populated) {
            this.scenario = scenario
            allEntities = ExploreMapSampleData.entities(scenario)
            _filters.value = ExploreMapSampleData.filters(scenario)
            when (scenario) {
                ExploreScenario.Loading -> _state.value = ExploreMapUiState.Loading
                ExploreScenario.Error -> _state.value = ExploreMapUiState.Error("Couldn't load the map.")
                else -> rebuild(selectedId = null)
            }
        }

        fun refresh() = load(scenario)

        // MARK: Type toggle

        fun selectKind(kind: ExploreKind?) {
            if (_activeKind.value == kind) return
            _activeKind.value = kind
            rebuild(selectedId = keptSelection())
        }

        // MARK: Sort

        fun selectSort(sort: ExploreSort) {
            if (_activeSort.value == sort) return
            _activeSort.value = sort
            rebuild(selectedId = currentSelectedId())
        }

        // MARK: Filters

        fun applyFilters(criteria: ExploreFilterCriteria) {
            _filters.value = criteria
            rebuild(selectedId = keptSelection())
        }

        /** Empty-state "Clear filters" — reset every dimension. */
        fun clearFilters() {
            _filters.value = ExploreFilterCriteria()
            _activeKind.value = null
            rebuild(selectedId = null)
        }

        /** Empty-state "Widen area" — open the radius to the widest stop. */
        fun widenArea() {
            _filters.value = _filters.value.copy(distanceUpper = ExploreFilterCriteria.DISTANCE_STOPS.last())
            rebuild(selectedId = currentSelectedId())
        }

        // MARK: Selection / sheet

        fun selectEntity(id: String?) {
            val current = _state.value as? ExploreMapUiState.Loaded ?: return
            _state.value = current.copy(selectedId = id)
        }

        fun setSheetStop(stop: ExploreSheetStop) {
            _sheetStop.value = stop
        }

        fun setClusterRadius(radiusDegrees: Double) {
            val clamped = radiusDegrees.coerceIn(0.0005, 0.05)
            if (kotlin.math.abs(clamped - clusterRadiusDegrees) < 1e-6) return
            clusterRadiusDegrees = clamped
            val current = _state.value as? ExploreMapUiState.Loaded ?: return
            rebuild(selectedId = current.selectedId)
        }

        // MARK: Projection

        private fun currentSelectedId(): String? = (_state.value as? ExploreMapUiState.Loaded)?.selectedId

        private fun keptSelection(): String? {
            val prior = currentSelectedId()
            return if (filtered().any { it.id == prior }) prior else null
        }

        private fun filtered(): List<ExploreEntity> =
            allEntities.filter { entity ->
                val kind = _activeKind.value
                (kind == null || entity.kind == kind) && _filters.value.matches(entity)
            }

        private fun sortFor(source: List<ExploreEntity>): List<ExploreEntity> =
            when (_activeSort.value) {
                ExploreSort.Closest -> source.sortedBy { it.distanceMiles }
                ExploreSort.Newest -> source
            }

        private fun rebuild(selectedId: String?) {
            val sorted = sortFor(filtered())
            _state.value =
                ExploreMapUiState.Loaded(
                    entities = sorted,
                    markers = cluster(sorted, clusterRadiusDegrees),
                    userCoordinate = _userCoordinate.value,
                    selectedId = selectedId,
                )
        }

        companion object {
            /** Grid-bucket clusterer — buckets of ≥2 collapse into a cluster. */
            fun cluster(
                entities: List<ExploreEntity>,
                radiusDegrees: Double,
            ): List<ExploreMarker> {
                if (radiusDegrees <= 0) return entities.map { ExploreMarker.Entity(it) }
                val buckets = linkedMapOf<String, MutableList<ExploreEntity>>()
                entities.forEach { entity ->
                    val key = bucketKey(entity.latitude, entity.longitude, radiusDegrees)
                    buckets.getOrPut(key) { mutableListOf() }.add(entity)
                }
                return buckets.entries.sortedBy { it.key }.map { (key, group) ->
                    if (group.size == 1) {
                        ExploreMarker.Entity(group[0])
                    } else {
                        val lats = group.map { it.latitude }
                        val lons = group.map { it.longitude }
                        val representative =
                            group.groupingBy { it.kind }.eachCount().maxByOrNull { it.value }?.key ?: group[0].kind
                        ExploreMarker.Cluster(
                            ExploreCluster(
                                id = key,
                                latitude = lats.average(),
                                longitude = lons.average(),
                                kind = representative,
                                count = group.size,
                                entityIds = group.map { it.id },
                                minLatitude = lats.min(),
                                maxLatitude = lats.max(),
                                minLongitude = lons.min(),
                                maxLongitude = lons.max(),
                            ),
                        )
                    }
                }
            }

            private fun bucketKey(
                latitude: Double,
                longitude: Double,
                radius: Double,
            ): String {
                val latBucket = kotlin.math.floor(latitude / radius).toInt()
                val lonBucket = kotlin.math.floor(longitude / radius).toInt()
                return "${latBucket}_$lonBucket"
            }
        }
    }
